src/main/java/com/google/devtools/build/lib/actions/FileContentsProxy.java

```diff
@@ -72,6 +94,20 @@ public boolean equals(Object other) {
     return ctime == that.ctime && nodeId == that.nodeId;
   }
 
+  /**
+   * Can be used when hardlink reference counter changes
+   * should not be considered a file modification.
+   * Is only comparing mtime and not ctime and is therefore
+   * not detecting changed metadata like permission.
+   */
+  public boolean isModified(FileContentsProxy other) {
+    if (other == this) {
+      return false;
+    }
+    // true if nodeId are different or inode has a new mtime
+    return nodeId != other.nodeId || mtime != other.mtime;
```

Comment:
Isn't this going to fall prey to the problems in #1525?

```
larsrc@mols:import-bazel-github-13279$ stat /tmp/a /tmp/b
  File: /tmp/a
  Size: 2         	Blocks: 8          IO Block: 4096   regular file
Device: fe01h/65025d	Inode: 2485767     Links: 1
Access: (0640/-rw-r-----)  Uid: (121084/  larsrc)   Gid: (89939/primarygroup)
Access: 2021-04-26 11:32:04.137402754 +0200
Modify: 2021-04-26 11:32:04.137402754 +0200
Change: 2021-04-26 11:32:04.137402754 +0200
 Birth: 2021-04-26 11:31:12.493443234 +0200
  File: /tmp/b
  Size: 2         	Blocks: 8          IO Block: 4096   regular file
Device: fe01h/65025d	Inode: 2485994     Links: 1
Access: (0640/-rw-r-----)  Uid: (121084/  larsrc)   Gid: (89939/primarygroup)
Access: 2021-04-26 11:32:04.137402754 +0200
Modify: 2021-04-26 11:32:04.137402754 +0200
Change: 2021-04-26 11:32:04.137402754 +0200
 Birth: 2021-04-26 11:31:12.493443234 +0200
larsrc@mols:import-bazel-github-13279$ mv /tmp/a /tmp/b
larsrc@mols:import-bazel-github-13279$ stat /tmp/a /tmp/b
stat: cannot statx '/tmp/a': No such file or directory
  File: /tmp/b
  Size: 2         	Blocks: 8          IO Block: 4096   regular file
Device: fe01h/65025d	Inode: 2485767     Links: 1
Access: (0640/-rw-r-----)  Uid: (121084/  larsrc)   Gid: (89939/primarygroup)
Access: 2021-04-26 11:32:04.137402754 +0200
Modify: 2021-04-26 11:32:04.137402754 +0200
Change: 2021-04-26 11:32:36.049377813 +0200
 Birth: 2021-04-26 11:31:12.493443234 +0200
```

Reply:
In your example, mtime is same after the overwrite of (b) but nodeId is different so we will be detecting this modification, if I understand your concern correctly.